player/js/utils/shapes/ZigZagModifier.js

```diff
@@ -0,0 +1,91 @@
+import {
+  extendPrototype,
+} from '../functionExtensions';
+import PropertyFactory from '../PropertyFactory';
+import shapePool from '../pooling/shape_pool';
+import {
+  ShapeModifier,
+} from './ShapeModifiers';
+import { PolynomialBezier } from '../PolynomialBezier';
+
+function ZigZagModifier() {}
+extendPrototype([ShapeModifier], ZigZagModifier);
+ZigZagModifier.prototype.initModifierProperties = function (elem, data) {
+  this.getValue = this.processKeys;
+  this.amplitude = PropertyFactory.getProp(elem, data.s, 0, null, this);
+  this.frequency = PropertyFactory.getProp(elem, data.pt, 0, null, this);
+  this._isAnimated = this.amplitude.effectsSequence.length !== 0 && this.frequency.effectsSequence.length !== 0;
+};
+
+ZigZagModifier.prototype.processPath = function (path, amplitude, frequency) {
+  var pathLength = path._length;
+  var clonedPath = shapePool.newElement();
+  clonedPath.c = path.c;
+  var vX;
+  var vY;
+  var direction = 1;
+
+  if (!path.c) {
+    pathLength -= 1;
+  }
+
+  for (var cur = 0; cur < pathLength; cur += 1) {
+    var bez = PolynomialBezier.shapeSegment(path, cur);
+
+    vX = bez.points[0][0];
+    vY = bez.points[0][1];
+    clonedPath.setTripleAt(vX, vY, vX, vY, vX, vY, clonedPath.length());
+
+    for (var i = 0; i < frequency; i += 1) {
+      var t = (i + 0.5) / frequency;
+      var pt = bez.point(t);
```

Comment:
For bezier curves with control points that don't match the vertices, this pt doesn't seem to be calculated right. It's most noticeable with curves with pronounced control point values.
I haven't done enough tests, but I'm pretty sure this is related to how AE splits the path in equal length segments instead of getting the "t" values at equidistant positions of the [0, 1] range.  
I've had to fix this in the past for trim path, and the solution is splitting the curve into multiple segments (in the order of hundreds) and find the right "t" for the corresponding value.  
This will take some time, and performance will degrade a little since calculating it is not that straightforward.  
I'm ok merging it the way it is and iterating over it on a following PR.

Reply:
Perhaps a way to optimize it could be to just split the bezier in a handful of segments and assume it's linear within those segments. This will introduce small errors in length calculations but it shouldn't be too noticeable (Based on some quick tests if you split each segments in 10 linear chunks you get ~2% error)